src/core.c/CompUnit/Repository/Installation.pm6

```diff
@@ -148,147 +148,200 @@ sub MAIN(:$name, :$auth, :$ver, *@, *%) {
+                  .spurt($dist.content($file).open(:bin).slurp(:close));
+            }
+
+            # resources/
+            for %files.kv -> $name-path, $file {
+                next unless $name-path.starts-with('resources/');
+
+                # $name-path is 'resources/libraries/p5helper' while
+                # $file is 'resources/libraries/libp5helper.so'
+                my $id := self!file-id(~$name-path, $dist-id)
+                  ~ '.'
+                  ~ $file.IO.extension;
+
+                %links{$name-path} := $id;
+                $resources-dir.add($id)
+                  .spurt($dist.content($file).open(:bin).slurp(:close));
+            }
+
+            my %meta = %($dist.meta);
+
+            # add our new name-path => content-id mapping
+            %meta<files> := %links;
+
+            # new meta data added to provides
+            %meta<provides> := %provides;
+
+            %!dist-metas{$dist-id} := %meta;
+            $dist-dir.add($dist-id).spurt:
+              Rakudo::Internals::JSON.to-json(%meta, :sorted-keys);
+
+            # reset cached id so it's generated again on next access.
+            # identity changes with every installation of a dist.
+            $!id = Any;
+
+            {  # make sure we have local $*DISTRIBUTION and $*RESOURCES
+
+                my $head := $*REPO;
+                # Precomp files should only depend on downstream repos
+                PROCESS::<$REPO> := self;
+
+                my $precomp     := $*REPO.precomp-repository;   # why not self?
+                my $repo-prefix := self!repo-prefix;
+
+                my $*DISTRIBUTION := CompUnit::Repository::Distribution.new(
+                  $dist, :repo(self), :$dist-id
+                );
+                my $*RESOURCES := Distribution::Resources.new(
+                  :repo(self), :$dist-id
                 );
-                %done{$id} = 1;
+
+                my $compiler-id := CompUnit::PrecompilationId.new-without-check(
+                  $*RAKU.compiler.id
+                );
+
+                my @provides-keys = %provides.keys.sort;
+
+                # clean out any existing
+                $precomp.store.delete(
+                  $compiler-id, 
+                  CompUnit::PrecompilationId.new-without-check(
+                    %provides{$_}.values.head<file>)
+                ) for @provides-keys;
+
+                my %done;
+                my @promises;
+                for @provides-keys -> $key {
+                    my $id := %provides{$key}.values.head<file>;
+                    if %done{$id} {
+                        note "(Already doing $id)" if $verbose;
+                        next;
+                    }
+
+                    my $source      := $sources-dir.add($id);
+                    my $source-file := $repo-prefix
+                      ?? $repo-prefix ~ $source.relative($.prefix)
+                      !! $source;
+
+                    @promises.push: start {
```

Comment:
Spawning a process for every module listed under provides is going to reach diminishing returns pretty quickly.